### PR TITLE
[hevce] Get SliceStructure caps from driver

### LINKED ISSUE #1832: [hevce] Get MaxSliceSize/SliceStructure caps from driver
Please port https://github.com/Intel-Media-SDK/MediaSDK/pull/1814 to HEVCe refactored code

### PR DESCRIPTION
**Pull request by yskuridi**
Fixes #1832

### CODE REVIEW

**Review by dmitryermilov — DISMISSED** [review]

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp [hunk 1]
```diff
@@ -430,6 +430,19 @@ mfxStatus DDI_VA::QueryCaps()
         m_priorityPar.m_MaxContextPriority = 0;
     }
 
+    if(AV(VAConfigAttribEncSliceStructure)!=VA_ATTRIB_NOT_SUPPORTED)
```
Comment (dmitryermilov) [inline _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:433]: could you add space after "if"

Comment (yskuridi) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp]: done

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp [hunk 2]
```diff
@@ -430,6 +430,19 @@ mfxStatus DDI_VA::QueryCaps()
...
     }
 
+    if (AV(VAConfigAttribEncSliceStructure)!=VA_ATTRIB_NOT_SUPPORTED)
+    {
+        // Attribute for VAConfigAttribEncSliceStructure includes both:
+        // (1) information about supported slice structure and
+        // (2) indication of support for max slice size feature
+        uint32_t sliceCapabilities = AV(VAConfigAttribEncSliceStructure);
+        uint32_t sliceStructure = sliceCapabilities & ~VA_ENC_SLICE_STRUCTURE_MAX_SLICE_SIZE;
+        m_caps.SliceStructure = ConvertSliceStructureVAAPIToMFX(sliceStructure);
```
Comment (lakulako) [inline _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:440]: I think it is better original sliceCapabilities to the function.

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp [hunk 3]
```diff
@@ -461,4 +474,21 @@ uint32_t DDI_VA::ConvertRateControlMFX2VAAPI(mfxU16 rateControl, bool bSWBRC)
     return 0;
 }
 
+uint32_t DDI_VA::ConvertSliceStructureVAAPIToMFX(uint32_t structure)
+{
+    switch (structure)
+    {
+        case VA_ENC_SLICE_STRUCTURE_POWER_OF_TWO_ROWS:
+            return uint32_t(1);
+        case VA_ENC_SLICE_STRUCTURE_EQUAL_ROWS:
```
Comment (lakulako) [inline _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:483]: I guess caps can return multiple supported structures in different bits. This function will not work with bit combinations.
No need to cast uint32_t(1)

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp [hunk 4]
```diff
@@ -461,4 +474,21 @@ uint32_t DDI_VA::ConvertRateControlMFX2VAAPI(mfxU16 rateControl, bool bSWBRC)
     return 0;
 }
 
+uint32_t DDI_VA::ConvertSliceStructureVAAPIToMFX(uint32_t structure)
+{
+    switch (structure)
+    {
+        case VA_ENC_SLICE_STRUCTURE_POWER_OF_TWO_ROWS:
+            return uint32_t(1);
```
Comment (vilichev) [inline _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:482]: Please avoid usage of 'magic numbers'

**Review by vilichev — CHANGES_REQUESTED** [review]

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp (hunk 4 above)
Comment (dmitryermilov) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:482]: Agree, Ideally we need to re-use approach from https://github.com/Intel-Media-SDK/MediaSDK/pull/1814/files . And ideally we need to change this hardcoded 1, 2,3, etc across all places

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp (hunk 3 above)
Comment (yskuridi) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:483]: > I guess caps can return multiple supported structures in different bits. This function will not work with bit combinations.

I see no hint in the code that multiple supported structures are considered on msdk side.

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp (hunk 4 above)
Comment (yskuridi) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp]: > Please avoid usage of 'magic numbers'

Done.

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp (hunk 3 above)
Comment (lakulako) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:483]: I mean in value from driver several bits can be set.

File: _studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_data.h [hunk 5]
```diff
@@ -968,6 +968,16 @@ namespace Base
...
 
+    /*SliceStructure indicates the restrictions set on the slice structure.*/
+    enum SliceStructure
+    {
+        ONESLICE = 0, // Once slice for the whole frame
+        ROW2ROW = 1, // Slices are power of 2 number of rows, all slices the same
+        ROWSLICE = 2, // Slices are any number of rows, all slices the same
+        ARBITRARY_ROW_SLICE = 3, // Slices are any number of rows, slices can be different
+        ARBITRARY_MB_SLICE = 4, // Slices are any number of MBs, slices can be different
+    };
```
Comment (dmitryermilov) [inline _studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_data.h:979]: @yskuridi , can we keep that awesome comments from _studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_legacy_defaults.cpp this in enum?

**Review by thomasli21801 — APPROVED** [review]

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp (hunk 3 above)
Comment (JeremyShang) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp:483]: I have same concern with with @lakulako,  driver caps actually can return multiple supported structure like this one. VA_ENC_SLICE_STRUCTURE_EQUAL_ROWS | VA_ENC_SLICE_STRUCTURE_MAX_SLICE_SIZE | VA_ENC_SLICE_STRUCTURE_EQUAL_MULTI_ROWS;

**Review by dmitryermilov — APPROVED** [review]

File: _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp (hunk 3 above)
Comment (dmitryermilov) [reply _studio/mfx_lib/encode_hw/hevc/linux/base/hevcehw_base_va_lin.cpp]: @aleonenk , @DenWolf , could you please also review ? I vagally recall you came across the same in AVCe

**Review by lakulako — APPROVED** [review]

**Review by vilichev — APPROVED** [review]

**Pull request merged**